components/script/dom/bindings/function.rs

```diff
@@ -0,0 +1,45 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#![allow(unsafe_code)]
+
+use std::ffi::c_char;
+use std::rc::Rc;
+
+use js::jsapi::{JSContext, JSObject, JS_GetFunctionObject, JS_NewFunction, Value};
+
+use crate::dom::bindings::codegen::Bindings::FunctionBinding::Function;
+use crate::dom::types::GlobalScope;
+
+type NativeFunction = unsafe extern "C" fn(*mut JSContext, u32, *mut Value) -> bool;
```

Comment:
This will still require user to write unsafe code. I wish it could take a normal rust function and we trampoline or wrap it in the method.

I'll look if there's some possible approach.

Reply:
Yes that sounds like a good idea, let us know what you find out.

Reply:
Hmm, it seems the extern C function can't capture the rust function, and there's no way to trampoline either.
How about providing a `native_fn` macro instead of `new_native` then:

- The signature should be `native_fn!(call, name, nargs, flags)`.
- `call` is a rust function which signature should be `fn(*mut JSContext, u32, CallArgs) -> bool`
- `name` should be `String` or `&str`. We will convert it to Cstr in the macro.

So the user can get the extern C function via `let c_fn: Rc<Function> = native_fn!(rust_fn, name, nargs, flags)` after they define their rust function.

Reply:
> name should be String or &str. We will convert it to Cstr in the macro.

I think that would require new alloc (as CStr is null-terminated, hence one byte more). Also using C-string literals gives Cstr: https://blog.rust-lang.org/2024/03/21/Rust-1.77.0.html#c-string-literals, so it should be easy enough to use.

Reply:
when I tried to call `call` from extern c wrapper I was getting:

```
can't capture dynamic environment in a fn item use the `|| { ... }` closure form instead
```

this how I could make it work :

```

#[macro_export]
macro_rules! native_fn {
    ($cx:expr, $rust_fn:expr, $name:expr, $nargs:expr, $flags:expr) => {{
        unsafe extern "C" fn wrapper(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
            (WRAPPER_FN.unwrap())(cx, argc, vp)
        }

        static mut WRAPPER_FN: Option<unsafe fn(*mut JSContext, u32, *mut JSVal) -> bool> = None;

        unsafe {
            WRAPPER_FN = Some($rust_fn);
            assert_eq!(*$name.last().unwrap(), b'\0');
            let raw_fun = crate::dom::bindings::import::module::jsapi::JS_NewFunction(
                *$cx,
                Some(wrapper),
                $nargs,
                $flags,
                $name.as_ptr() as *const std::ffi::c_char,
            );
            assert!(!raw_fun.is_null());
            crate::dom::bindings::import::module::jsapi::JS_GetFunctionObject(raw_fun)
        }
    }};
}
```

Reply:
@wusyong what do you think?

Reply:
It seems fine when I test this:
```rust
#[macro_export]
macro_rules! native_raw_obj_fn {
    ($cx:expr, $call:expr, $name:expr, $nargs:expr, $flags:expr) => {{
        unsafe extern "C" fn wrapper(cx: *mut JSContext, argc: u32, vp: *mut JSVal) -> bool {
            $call(cx, argc, vp)
        }

        unsafe {
            assert_eq!(*$name.last().unwrap(), b'\0');
            let raw_fun = crate::dom::bindings::import::module::jsapi::JS_NewFunction(
                *$cx,
                Some(wrapper),
                $nargs,
                $flags,
                $name.as_ptr() as *const std::ffi::c_char,
            );
            assert!(!raw_fun.is_null());
            crate::dom::bindings::import::module::jsapi::JS_GetFunctionObject(raw_fun)
        }
    }};
}
```

Which case did you have trouble with? Count and ByteLenghtQueuing both can work in my test.

Reply:
Indeed, this works. I'm not sure what went wrong with my previous example.